src/data_structures/fmindex.rs

```diff
@@ -88,41 +100,78 @@ pub trait FMIndexable {
 
 }
 
-/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
-/// intervals matching a given pattern.
+pub trait FMIndexCore {
+    type SA: SuffixArray;
 
-#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
-pub struct FMIndex<
-        DBWT: DerefBWT + Clone,
-        DLess: DerefLess + Clone,
-        DOcc: DerefOcc + Clone> {
-    bwt: DBWT,
-    less: DLess,
-    occ: DOcc,
+    fn occ(&self, r: usize, a: u8) -> usize;
+    fn less(&self, a: u8) -> usize;
+    fn bwt(&self) -> &BWT;
+    fn sa(&self) -> &Self::SA;
 }
 
-impl<
-    DBWT: DerefBWT + Clone,
-    DLess: DerefLess + Clone,
-    DOcc: DerefOcc + Clone> FMIndexable for FMIndex<DBWT, DLess, DOcc> {
+impl FMIndexCore for SampledSuffixArray {
+    // the trait needs to track what kind of suffix array reference it's returning
+    type SA = SampledSuffixArray;
 
     fn occ(&self, r: usize, a: u8) -> usize {
-        self.occ.get(&self.bwt, r, a)
+        self.occ(r, a)
     }
+
     fn less(&self, a: u8) -> usize {
-        self.less[a as usize]
+        self.less(a)
     }
-    /// Provide a reference to the underlying BWT.
+
     fn bwt(&self) -> &BWT {
-        &self.bwt
+        self.bwt()
+    }
+
+    fn sa(&self) -> &Self::SA {
+        &self
+    }
+}
+
+impl FMIndexCore for (RawSuffixArray, BWT, Occ, Less) {
```

Comment:
I recall that @johanneskoester wanted to keep the suffix array container separate from the FM-Index so that users could perform searches and produce intervals separately from needing to hold onto the suffix array.